x/supplier/keeper/msg_server_unstake_supplier_test.go

```diff
@@ -54,19 +36,87 @@ func TestMsgServer_UnstakeSupplier_Success(t *testing.T) {
 	foundSupplier, isSupplierFound := k.GetSupplier(ctx, supplierAddr)
 	require.True(t, isSupplierFound)
 	require.Equal(t, supplierAddr, foundSupplier.Address)
-	require.Equal(t, initialStake.Amount, foundSupplier.Stake.Amount)
+	require.Equal(t, math.NewInt(initialStake), foundSupplier.Stake.Amount)
 	require.Len(t, foundSupplier.Services, 1)
 
-	// Unstake the supplier
+	// Initiate the supplier unstaking
 	unstakeMsg := &types.MsgUnstakeSupplier{Address: supplierAddr}
 	_, err = srv.UnstakeSupplier(ctx, unstakeMsg)
 	require.NoError(t, err)
 
-	// Make sure the supplier can no longer be found after unstaking
-	_, isSupplierFound = k.GetSupplier(ctx, supplierAddr)
+	// Make sure the supplier entered the unbonding period
+	foundSupplier, isSupplierFound = k.GetSupplier(ctx, supplierAddr)
+	require.True(t, isSupplierFound)
+	require.Greater(t, foundSupplier.UnstakeCommitSessionEndHeight, int64(0))
+
+	// Move block height to the end of the unbonding period
+	unbondingHeight := foundSupplier.UnstakeCommitSessionEndHeight + int64(k.GetParams(ctx).SupplierUnbondingPeriodBlocks)
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	sdkCtx = sdkCtx.WithBlockHeight(unbondingHeight)
+
+	// Run the endblocker to unbond the supplier
+	err = k.EndBlockerUnbondSupplier(sdkCtx)
+	require.NoError(t, err)
+
+	// Make sure the supplier is removed from the suppliers list when the
+	// unbonding period is over
+	_, isSupplierFound = k.GetSupplier(sdkCtx, supplierAddr)
 	require.False(t, isSupplierFound)
 }
 
+func TestMsgServer_UnstakeSupplier_CancelUnbondingIfRestaked(t *testing.T) {
+	k, ctx := keepertest.SupplierKeeper(t)
+	srv := keeper.NewMsgServerImpl(k)
+
+	// Generate an address for the supplier
+	supplierAddr := sample.AccAddress()
+
+	// Stake the supplier
+	initialStake := int64(100)
+	stakeMsg := createStakeMsg(supplierAddr, initialStake)
+	_, err := srv.StakeSupplier(ctx, stakeMsg)
+	require.NoError(t, err)
+
+	// Verify that the supplier exists with no unbonding height
+	foundSupplier, isSupplierFound := k.GetSupplier(ctx, supplierAddr)
+	require.True(t, isSupplierFound)
+	require.Equal(t, int64(0), foundSupplier.UnstakeCommitSessionEndHeight)
+
+	// Initiate the supplier unstaking
+	unstakeMsg := &types.MsgUnstakeSupplier{Address: supplierAddr}
+	_, err = srv.UnstakeSupplier(ctx, unstakeMsg)
+	require.NoError(t, err)
+
+	// Make sure the supplier entered the unbonding period
+	foundSupplier, isSupplierFound = k.GetSupplier(ctx, supplierAddr)
+	require.True(t, isSupplierFound)
+	require.Greater(t, foundSupplier.UnstakeCommitSessionEndHeight, int64(0))
+
+	unbondingHeight := foundSupplier.UnstakeCommitSessionEndHeight + int64(k.GetParams(ctx).SupplierUnbondingPeriodBlocks)
+
+	// Stake the supplier again
+	stakeMsg = createStakeMsg(supplierAddr, initialStake+1)
```

Comment:
Yes it is to avoid hitting the "Same staking amount" error.

I don't fully grasp why upstaking is needed. But if it's to avoid spamming/quickly-changing the supplier info, then we should keep it for staking cancellation too.

Otherwise we'd circumvent the upstaking by allowing suppliers to unstake->cancel (unless we check that they did not change any param).